src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,397 @@
+            right = memory.Slice(ichCur + 1, memory.Length - ichCur - 1);
+            return true;
+        }
+
+        /// <summary>
+        /// Splits <paramref name="memory"/> on the left-most occurrence of an element of separators character array and
+        /// produces the left and right <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> values. If <paramref name="memory"/> does not contain any of the
+        /// characters in separators, this return false and initializes <paramref name="left"/> to this instance
+        /// and <paramref name="right"/> to the default <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> value.
+        /// </summary>
+        public static bool SplitOne(char[] separators, out ReadOnlyMemory<char> left, out ReadOnlyMemory<char> right, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValueOrNull(separators);
+
+            if (memory.IsEmpty || separators == null || separators.Length == 0)
+            {
+                left = memory;
+                right = default;
+                return false;
+            }
+
+            var span = memory.Span;
+
+            int ichCur = 0;
+            if (separators.Length == 1)
+            {
+                // Note: This duplicates code of the other SplitOne, but doing so improves perf because this is
+                // used so heavily in instances parsing.
+                char chSep = separators[0];
+                for (; ; ichCur++)
+                {
+                    Contracts.Assert(0 <= ichCur && ichCur <= memory.Length);
+                    if (ichCur >= memory.Length)
+                    {
+                        left = memory;
+                        right = default;
+                        return false;
+                    }
+                    if (span[ichCur] == chSep)
+                        break;
+                }
+            }
+            else
+            {
+                for (; ; ichCur++)
+                {
+                    Contracts.Assert(0 <= ichCur && ichCur <= memory.Length);
+                    if (ichCur >= memory.Length)
+                    {
+                        left = memory;
+                        right = default;
+                        return false;
+                    }
+                    // REVIEW: Can this be faster?
+                    if (ContainsChar(span[ichCur], separators))
+                        break;
+                }
+            }
+
+            left = memory.Slice(0, ichCur);
+            right = memory.Slice(ichCur + 1, memory.Length - ichCur - 1);
+            return true;
+        }
+
+        /// <summary>
+        /// Returns a <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> with leading and trailing spaces trimmed. Note that this
+        /// will remove only spaces, not any form of whitespace.
+        /// </summary>
+        public static ReadOnlyMemory<char> Trim(ReadOnlyMemory<char> memory)
```

Comment:
I kind of think we shouldn't have a public method named `Trim` that performs differently than the typical .NET "Trim()" method.  Trim will trim any whitespace, not just spaces.
I think we should either rename the methods, or make them internal.

Reply:
Let's rename at least and make internal if we can. I'm very curious however to see what application requires such a seemingly oddly specifically different semantics

Reply:
Renamed but internal will throw error due to protection level.

---
In reply to: [216758599](https://github.com/dotnet/machinelearning/pull/863#discussion_r216758599) [](ancestors = 216758599)